pkg/api/config.go

```diff
@@ -3,71 +3,177 @@ package api
+		} else if typeCount == 1 {
+			// TODO remove when the migration is completed
+			if len(test.From) > 0 {
+				validationErrors = append(validationErrors, fmt.Errorf("%s[%d] specifies both `From` and a test type", fieldRoot, num))
 			}
+		} else if typeCount > 1 {
+			validationErrors = append(validationErrors, fmt.Errorf("%s[%d] has more than one type", fieldRoot, num))
 		}
 	}
-	buildRootImage := config.InputConfiguration.BuildRootImage
-	if buildRootImage != nil {
-		if buildRootImage.ProjectImageBuild != nil && buildRootImage.ImageStreamTagReference != nil {
-			errs = append(errs, errors.New("both git_source_image and image_stream_tag cannot be set for the build_root"))
-		} else if buildRootImage.ProjectImageBuild == nil && buildRootImage.ImageStreamTagReference == nil {
-			errs = append(errs, errors.New("you have to specify either git_source_image or image_stream_tag for the build_root"))
-		}
+	return kerrors.NewAggregate(validationErrors)
+}
+
+func validateImageStreamTagReference(fieldRoot string, input ImageStreamTagReference) error {
+	var validationErrors []error
+
+	if _, err := url.ParseRequestURI(input.Cluster); err != nil && input.Cluster != "" {
+		validationErrors = append(validationErrors, fmt.Errorf("%s.cluster invalid URL given: %s", fieldRoot, err))
+	}
+
+	if input.Tag == "" {
+		validationErrors = append(validationErrors, fmt.Errorf("%s.tag: value required but not provided", fieldRoot))
+	}
+
+	return kerrors.NewAggregate(validationErrors)
+}
+
+func validateImageStreamTagReferenceMap(fieldRoot string, input map[string]ImageStreamTagReference) error {
+	var validationErrors []error
+	// map has to be sorted to avoid flakes in tests.
+	var keys []string
+	for k := range input {
+		keys = append(keys, k)
+	}
+
+	sort.Strings(keys)
+	for _, k := range keys {
+		validationErrors = append(validationErrors, validateImageStreamTagReference(fmt.Sprintf("%s.%s", fieldRoot, k), input[k]))
 	}
-	return kerrors.NewAggregate(errs)
+	return kerrors.NewAggregate(validationErrors)
 }
 
-func validateTargetCloud(test string, tc TargetCloud) error {
+func validatePromotionConfiguration(fieldRoot string, input PromotionConfiguration) error {
+	if len(input.Name) == 0 && len(input.Tag) == 0 {
+		return fmt.Errorf("%s: no name or tag defined for promotion config", fieldRoot)
+	}
+	return nil
+}
+
+func validateReleaseTagConfiguration(fieldRoot string, input ReleaseTagConfiguration) error {
+	if len(input.Name) == 0 && len(input.Tag) == 0 {
+		return fmt.Errorf("%s: no name or tag defined for promotion config", fieldRoot)
+	}
+	return nil
+}
+
+func validateTargetCloud(fieldRoot string, tc TargetCloud) error {
 	if tc != TargetCloudAWS && tc != TargetCloudGCP {
-		return fmt.Errorf("test %q: invalid target cloud %q", test, tc)
+		return fmt.Errorf("%q: invalid target cloud %q", fieldRoot, tc)
+	}
+	return nil
+}
+
+func searchForTestDuplicates(tests []TestStepConfiguration) error {
```

Comment:
We also need to ensure that there are no collisions between:
 - input base images
 - input base RPM images
 - build root (users can't set this tag name but we need to make sure they don't set any of the other inputs to be it)

Reply:
Is there a reason for `tests` to be an array instead of a map?

Reply:
We would need a follow-up issue for that.